core/shared/src/main/scala/zio/FiberRef.scala

```diff
@@ -52,12 +52,49 @@ import zio.stacktracer.TracingImplicits.disableAutoTrace
  * Here `value` will be 2 as the value in the joined fiber is lower and we
  * specified `max` as our combining function.
  */
-final class FiberRef[A] private[zio] (
-  private[zio] val initial: A,
-  private[zio] val fork: A => A,
-  private[zio] val join: (A, A) => A
-) extends Serializable { self =>
-  type ValueType = A
+trait FiberRef[A] extends Serializable { self =>
+
+  /**
+   * The type of the value of the `FiberRef`.
+   */
+  type Value = A
+
+  /**
+   * The type of the patch that describes updates to the value of the
+   * `FiberRef`. In the simple case this will just be a function that sets the
+   * value of the `FiberRef`. In more complex cases this will describe an update
+   * to a piece of a whole value, allowing updates to the value by different
+   * fibers to be combined in a compositional way when those fibers are joined.
+   */
+  type Patch
+
+  /**
+   * The initial value of the `FiberRef`.
+   */
+  def initial: Value
+
+  /**
+   * Constructs a patch describing the updates to a value from an old value and
+   * a new value.
+   */
+  def diff(oldValue: Value, newValue: Value): Patch
+
+  /**
+   * Combines two patches to produce a new patch that describes the updates of
+   * the first patch and then the updates of the second patch.
+   */
+  def combine(first: Patch, second: Patch): Patch
```

Comment:
We should state that the operation should be associative, and also observe further that if it is commutative, then fiber refs will behave better with respect to concurrent operations.